src/H5Tconv.c

```diff
@@ -2966,18 +3079,18 @@ H5T__conv_enum(H5T_t *src, H5T_t *dst, H5T_cdata_t *cdata, const H5T_conv_ctx_t
                 src_delta = dst_delta = (ssize_t)buf_stride;
                 s = d = buf;
             }
-            else if (dst->shared->size <= src->shared->size) {
-                H5_CHECKED_ASSIGN(src_delta, ssize_t, src->shared->size, size_t);
-                H5_CHECKED_ASSIGN(dst_delta, ssize_t, dst->shared->size, size_t);
+            else if (dst_sh->size <= src_sh->size) {
+                H5_CHECKED_ASSIGN(src_delta, ssize_t, src_sh->size, size_t);
+                H5_CHECKED_ASSIGN(dst_delta, ssize_t, dst_sh->size, size_t);
                 s = d = buf;
             }
             else {
-                H5_CHECK_OVERFLOW(src->shared->size, size_t, ssize_t);
-                H5_CHECK_OVERFLOW(dst->shared->size, size_t, ssize_t);
-                src_delta = -(ssize_t)src->shared->size;
-                dst_delta = -(ssize_t)dst->shared->size;
-                s         = buf + (nelmts - 1) * src->shared->size;
-                d         = buf + (nelmts - 1) * dst->shared->size;
+                H5_CHECK_OVERFLOW(src_sh->size, size_t, ssize_t);
+                H5_CHECK_OVERFLOW(dst_sh->size, size_t, ssize_t);
+                src_delta = -(ssize_t)src_sh->size;
+                dst_delta = -(ssize_t)dst_sh->size;
+                s         = buf + (nelmts - 1) * src_sh->size;
+                d         = buf + (nelmts - 1) * dst_sh->size;
             }
 
             for (i = 0; i < nelmts; i++, s += src_delta, d += dst_delta) {
```

Comment:
Small optimization:  I suggest hoisting the if/else outside of the loop (duplicating the loop in each branch), since 'priv->length' is constant for all elements.

Reply:
I'd be surprised if most compilers don't optimize that with branch prediction, but I can do that.